fvm/evm/emulator/state/collection.go

```diff
@@ -0,0 +1,422 @@
+func (cp *CollectionProvider) Commit() error {
+	return cp.storage.FastCommit(runtime.NumCPU())
+}
+
+// Collection provides a persistent and compact way of storing key/value pairs
+// each collection has a unique collectionID that can be used to fetch the collection
+//
+// TODO(ramtin): we might not need any extra hashing on the atree side
+// and optimize this to just use the key given the keys are hashed ?
+type Collection struct {
+	omap         *atree.OrderedMap
+	storage      *atree.PersistentSlabStorage
+	collectionID []byte
+}
+
+// CollectionID returns the unique id for the collection
+func (c *Collection) CollectionID() []byte {
+	return c.collectionID
+}
+
+// Get gets the value for the given key
+//
+// if key doesn't exist it returns nil (no error)
+func (c *Collection) Get(key []byte) ([]byte, error) {
+	// first check if we have the key
+	// this avoids getting a NotFound error
+	found, err := c.omap.Has(compare, hashInputProvider, NewByteStringValue(key))
+	if err != nil {
+		return nil, err
+	}
+	if !found {
+		return nil, nil
+	}
+
+	data, err := c.omap.Get(compare, hashInputProvider, NewByteStringValue(key))
+	if err != nil {
+		return nil, err
+	}
+
+	value, err := data.StoredValue(c.omap.Storage)
+	if err != nil {
+		return nil, err
+	}
+
+	return value.(ByteStringValue).Bytes(), nil
+}
+
+// Set sets the value for the given key
+//
+// if a value already stored at the given key it replaces the value
+func (c *Collection) Set(key, value []byte) error {
+	existingValueStorable, err := c.omap.Set(compare, hashInputProvider, NewByteStringValue(key), NewByteStringValue(value))
+	if err != nil {
+		return err
+	}
+
+	if id, ok := existingValueStorable.(atree.StorageIDStorable); ok {
+		// NOTE: deep remove isn't necessary because value is ByteStringValue (not container)
+		err := c.storage.Remove(atree.StorageID(id))
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Remove removes a key from the collection
+//
+// if the key doesn't exist it return no error
+func (c *Collection) Remove(key []byte) error {
+	found, err := c.omap.Has(compare, hashInputProvider, NewByteStringValue(key))
+	if err != nil {
+		return err
+	}
+	if !found {
+		return nil
+	}
+
+	_, existingValueStorable, err := c.omap.Remove(compare, hashInputProvider, NewByteStringValue(key))
+	if err != nil {
+		return err
+	}
+
+	if id, ok := existingValueStorable.(atree.StorageIDStorable); ok {
+		// NOTE: deep remove isn't necessary because value is ByteStringValue (not container)
+		err := c.storage.Remove(atree.StorageID(id))
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Destroy destroys the whole collection
+func (c *Collection) Destroy() error {
+	var cachedErr error
+	err := c.omap.PopIterate(func(_ atree.Storable, valueStorable atree.Storable) {
+		if id, ok := valueStorable.(atree.StorageIDStorable); ok {
+			err := c.storage.Remove(atree.StorageID(id))
+			if err != nil && cachedErr == nil {
+				cachedErr = err
```

Comment:
is it ok for later errors to be ignored?

Reply:
Yeah, doesn't matter which error we return usually the first one is sufficient.